File: python/perfect-numbers/perfect_numbers.py
def classify(number):
    if number <= 0:
        raise ValueError

    aliquot_sum = compute_aliquot_sum(number)
    if aliquot_sum == number:
        return 'perfect'
    elif aliquot_sum > number:
        return 'abundant'
    else:
        return 'deficient'


def compute_aliquot_sum(n):
    if n == 1:
        return 0

    result = 1
    i = 2
    while i * i <= n:
        if n % i == 0:
            result += i
            if i * i != n:
                result += n // i
        i += 1
    return result

File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers
  ( classify
  , Classification(..)
  ) where

data Classification
  = Deficient
  | Perfect
  | Abundant
  deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify number =
  if number <= 0
    then Nothing
    else let aliquotSum = computeAliquotSum number
          in case (aliquotSum `compare` number) of
               EQ -> Just Perfect
               GT -> Just Abundant
               LT -> Just Deficient

computeAliquotSum :: Int -> Int
computeAliquotSum 1 = 0
computeAliquotSum n =
  1 +
  (sum $
   map
     (\i ->
        if i * i == n
          then i
          else i + n `div` i) $
   filter (\i -> n `mod` i == 0) $ takeWhile (\i -> i * i <= n) [2 ..])
